### PR TITLE
[DAL] Add task to template parameters for all algos

### PR DESCRIPTION
**Pull request by rlnx**
Changes:
- Port all API definitions from raw RST to the "Tool"
- Add missing sections about inference
- Add missing examples 
- Add `Task` to template parameters for all algorithms

**Links:**
- K-Means: https://rlnx.github.io/oneapi-spec/task-api/algorithms/clustering/kmeans.html
- kNN: https://rlnx.github.io/oneapi-spec/task-api/algorithms/nearest_neighbors/knn_classification.html
- PCA: https://rlnx.github.io/oneapi-spec/task-api/algorithms/decomposition/pca.html

### CODE REVIEW

File: source/elements/oneDAL/include/oneapi/dal/pca.hpp [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+   /// Tag-type that denotes `brute-force <pca_t_math_cov>`_ computational
+   /// method.
+   struct cov {};
+
+   /// Tag-type that denotes `k-d tree <pca_t_math_svd>`_ computational method.
+   struct svd {};
+
+   /// Alias tag-type for `brute-force <pca_t_math_cov>`_ computational
+   /// method.
+   using by_default = cov;
```
Comment (michael-smirnov) [inline source/elements/oneDAL/include/oneapi/dal/pca.hpp:17]: Brute-force, k-d tree - seems like doxygen comments were copied from KNN. References also don't work

File: source/elements/oneDAL/source/algorithms/decomposition/pca.rst [hunk 2]
```diff
@@ -68,397 +93,105 @@ signum function,
...
+Usage example
+-------------
+Training
+--------
+.. onedal_code:: oneapi::dal::pca::example::run_training
 
-   Alias tag-type for the `Covariance-based method`_.
+Inference
+---------
+.. onedal_code:: oneapi::dal::pca::example::run_inference
```
Comment (michael-smirnov) [inline source/elements/oneDAL/source/algorithms/decomposition/pca.rst:133]: You are using `print_table()` service function. Is it documented somewhere?

File: source/elements/oneDAL/source/algorithms/decomposition/pca.rst [hunk 3]
```diff
@@ -68,397 +93,105 @@ signum function,
...
 
+---------------------
+Programming Interface
+---------------------
```
Comment (michael-smirnov) [inline source/elements/oneDAL/source/algorithms/decomposition/pca.rst:137]: Do you want to add the sentence about namespace/header file in this PR or in some other?

File: source/elements/oneDAL/source/algorithms/clustering/kmeans.rst [hunk 4]
```diff
@@ -9,6 +9,33 @@ The K-Means algorithm partitions :math:`n` feature vectors into :math:`k`
 *clusters* minimizing some criterion. Each cluster is characterized by a
```
Comment (michael-smirnov) [inline source/elements/oneDAL/source/algorithms/clustering/kmeans.rst:9]: Don't you want to connect "clusters" with "Clustering" task in the glossary or create its own glossary term?

File: source/elements/oneDAL/source/algorithms/nearest_neighbors/knn_classification.rst [hunk 5]
```diff
@@ -127,36 +123,54 @@ distance between :math:`x_j'` and the most distant feature vector from
...
+
+Inference
+---------
+.. onedal_code:: oneapi::dal::knn::example::run_inference
+
 
 ---------------------
 Programming Interface
 ---------------------
+All type and function in this section shall be declared in the
```
Comment (outoftardis) [inline source/elements/oneDAL/source/algorithms/nearest_neighbors/knn_classification.rst:141]:
```suggestion
All types and functions in this section shall be declared in the
```

File: source/elements/oneDAL/source/glossary.rst [hunk 6]
```diff
@@ -48,6 +48,11 @@ Machine learning terms
     Dataset
         A collection of :capterm:`observations <observation>`.
 
+    Dimensionality reduction
+        A problem of transforming a set of :capterm:`feature vectors <feature
+        vector>` from a high-dimensional space into a low-dimensional space
+        retaining meaningful properties of the original feature vectors
```
Comment (outoftardis) [inline source/elements/oneDAL/source/glossary.rst:54]:
```suggestion
        while retaining meaningful properties of the original feature vectors.
```

File: source/elements/oneDAL/include/oneapi/dal/pca.hpp [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+/// @tparam Task   Tag-type that specifies type of the problem to solve. Can
+///                be :expr:`task::dim_reduction`.
+///
+/// @param[in] desc  Descriptor of the algorithm
+/// @param[in] input Input data for the inference operation
+/// @return result   Result of the inference operation
+///
+/// @pre :expr:`input.data.has_data == true`
+/// @pre :expr:`input.model.eigenvectors.row_count == desc.component_count`
+/// @pre :expr:`input.model.eigenvectors.column_count = input.data.column_count`
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/pca.hpp:210]:
```suggestion
/// @pre :expr:`input.model.eigenvectors.column_count == input.data.column_count`
```

File: source/elements/oneDAL/include/oneapi/dal/pca.hpp [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+
+   /// The number of principal components $r$. If it is zero, the algorithm
+   /// computes the eigenvectors for all features, $r = p$.
+   /// @remark default = 0
+   /// @invariant :expr:`component_count >= 0`
+   int64_t get_component_count() const;
+   descriptor& set_component_count(int64_t);
+
+   /// Specifies whether the algorithm applies the `Sign-flip technique`_.
+   /// If it is `true`, directions of the eigenvectors must be deterministic.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/pca.hpp:53]:
```suggestion
   /// If it is `true`, the directions of the eigenvectors must be deterministic.
```

File: source/elements/oneDAL/include/oneapi/dal/pca.hpp [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+/// @tparam Task Tag-type that specifies type of the problem to solve. Can
+///              be :expr:`task::dim_reduction`.
+template <typename Task = task::by_default>
+class model {
+public:
+   /// Creates a new instance of the class with the default property values.
+   model();
+
+   /// $r \\times p$ table with the eigenvectors. Each row contains one
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/pca.hpp:67]:
```suggestion
   /// An $r \\times p$ table with the eigenvectors. Each row contains one
```

File: source/elements/oneDAL/include/oneapi/dal/kmeans.hpp [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+   /// Alias tag-type for custering task.
+   using by_default = clustering;
+} // namespace task
+
+/// @tparam Float  The floating-point type that the algorithm uses for
+///                intermediate computations. Can be :expr:`float` or
+///                :expr:`double`.
+/// @tparam Method Tag-type that specifies an implementation of algorithm. Can
+///                be :expr:`method::lloyd`.
+/// @tparam Task   Tag-type that specifies type of the problem to solve. Can
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans.hpp:31]:
```suggestion
/// @tparam Task   Tag-type that specifies the type of the problem to solve. Can
```

apply this everywhere else as well

File: source/elements/oneDAL/include/oneapi/dal/pca.hpp [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+/// @tparam Task Tag-type that specifies type of the problem to solve. Can
+///              be :expr:`task::dim_reduction`.
+template <typename Task = task::by_default>
+class train_input {
+public:
+   /// Creates a new instance of the class with the given :literal:`data`
+   /// property value
+   train_input(const table& data = table{});
+
+   /// $n \\times p$ table with the training data, where each row stores one
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/pca.hpp:87]:
```suggestion
   /// An $n \\times p$ table with the training data, where each row stores one
```

File: source/elements/oneDAL/include/oneapi/dal/pca.hpp [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+class train_result {
+public:
+   /// Creates a new instance of the class with the default property values.
+   train_result();
+
+   /// The trained PCA model
+   /// @remark default = model<Task>{}
+   const model<Task>& get_model() const;
+
+   /// $1 \\times r$ table that contains mean value for the first $r$
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/pca.hpp:106]:
```suggestion
   /// A $1 \\times r$ table that contains the mean values for the first $r$
```

File: source/elements/oneDAL/include/oneapi/dal/pca.hpp [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+   /// features.
+   /// @remark default = table{}
+   const table& get_means() const;
+
+   /// $1 \\times r$ table that contains variance for the first $r$
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/pca.hpp:111]:
```suggestion
   /// A $1 \\times r$ table that contains the variances for the first $r$
```

File: source/elements/oneDAL/include/oneapi/dal/pca.hpp [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+   /// features.
+   /// @remark default = table{}
+   const table& get_variances() const;
+
+   /// $1 \\times r$ table that contains eigenvalue for for the first
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/pca.hpp:116]:
```suggestion
   /// A $1 \\times r$ table that contains the eigenvalues for for the first
```

File: source/elements/oneDAL/include/oneapi/dal/pca.hpp [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+   /// $r$ features.
+   /// @remark default = table{}
+   const table& get_eigenvalues() const;
+
+   /// $r \\times p$ table with the eigenvectors. Each row contains one
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/pca.hpp:121]:
```suggestion
   /// An $r \\times p$ table with the eigenvectors. Each row contains one
```

File: source/elements/oneDAL/include/oneapi/dal/pca.hpp [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+   /// eigenvector.
+   /// @remark default = table{}
+   /// @invariant :expr:`eigenvectors == model.eigenvectors`
+   const table& get_eigenvectors() const;
+};
+
+/// Runs the training operation for PCA. For more details see
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/pca.hpp:128]:
```suggestion
/// Runs the training operation for PCA. For more details, see
```

File: source/elements/oneDAL/source/algorithms/decomposition/pca.rst [hunk 17]
```diff
@@ -68,397 +86,108 @@ signum function,
...
+.. _pca_i_math:
 
-      const auto pca_desc = oneapi::dal::pca::desc<float>{}
-         .set_component_count(model.get_component_count());
+Inference
+---------
+Given the inference set :math:`X' = \{ x_1', \ldots, x_m' \}` of
+:math:`p`-dimensional feature vectors and the :math:`r \times p` matrix
+:math:`T` produced at the training stage, the problem is to transform :math:`X'`
+to the set :math:`X'' = \{ x_1'', \ldots, x_m'' \}`, where :math:`x_{j}''` is
```
Comment (outoftardis) [inline source/elements/oneDAL/source/algorithms/decomposition/pca.rst:96]:
```suggestion
to the set :math:`X'' = \{ x_1'', \ldots, x_m'' \}`, where :math:`x_{j}''` is an
```

File: source/elements/oneDAL/source/algorithms/decomposition/pca.rst [hunk 18]
```diff
@@ -68,397 +86,108 @@ signum function,
...
+:math:`r`-dimensional feature vector, :math:`1 \leq j \leq m`.
 
-      const auto result = oneapi::dal::infer(pca_desc, model, new_data);
+The feature vector :math:`x_{j}''` is computed by applying linear transformation
```
Comment (outoftardis) [inline source/elements/oneDAL/source/algorithms/decomposition/pca.rst:99]:
```suggestion
The feature vector :math:`x_{j}''` is computed through applying linear transformation
```

File: source/elements/oneDAL/source/algorithms/decomposition/pca.rst [hunk 19]
```diff
@@ -68,397 +86,108 @@ signum function,
...
 
-   Alias tag-type for the `Covariance-based method`_.
+Inference
+---------
+.. onedal_code:: oneapi::dal::pca::example::run_inference
 
+---------------------
+Programming Interface
+---------------------
+All type and function in this section shall be declared in the
```
Comment (outoftardis) [inline source/elements/oneDAL/source/algorithms/decomposition/pca.rst:131]:
```suggestion
All types and functions in this section shall be declared in the
```

File: source/elements/oneDAL/source/algorithms/decomposition/pca.rst [hunk 20]
```diff
@@ -68,397 +86,108 @@ signum function,
...
+---------
+Given the inference set :math:`X' = \{ x_1', \ldots, x_m' \}` of
+:math:`p`-dimensional feature vectors and the :math:`r \times p` matrix
+:math:`T` produced at the training stage, the problem is to transform :math:`X'`
+to the set :math:`X'' = \{ x_1'', \ldots, x_m'' \}`, where :math:`x_{j}''` is
+:math:`r`-dimensional feature vector, :math:`1 \leq j \leq m`.
 
-      const auto result = oneapi::dal::infer(pca_desc, model, new_data);
+The feature vector :math:`x_{j}''` is computed by applying linear transformation
+[Lang87]_ with the matrix :math:`T` to the feature vector :math:`x_{j}'`,
```
Comment (outoftardis) [inline source/elements/oneDAL/source/algorithms/decomposition/pca.rst:100]: the phrasing  "with the matrix :math:`T`" is not clear

File: source/elements/oneDAL/source/algorithms/clustering/kmeans.rst [hunk 21]
```diff
@@ -69,430 +95,102 @@ The steps (1) and (2) are performed until the following **stop condition**,
...
 
-   Alias tag-type for the `Lloyd's method`_.
+Inference
+---------
+.. onedal_code:: oneapi::dal::kmeans::example::run_inference
 
+---------------------
+Programming Interface
+---------------------
+All type and function in this section shall be declared in the
```
Comment (outoftardis) [inline source/elements/oneDAL/source/algorithms/clustering/kmeans.rst:134]:
```suggestion
All types and functions in this section shall be declared in the
```

File: source/elements/oneDAL/source/algorithms/clustering/kmeans.rst [hunk 22]
```diff
@@ -69,430 +95,102 @@ The steps (1) and (2) are performed until the following **stop condition**,
...
-      struct lloyd {};
-      using by_default = lloyd;
-   } // namespace method
+.. _kmeans_i_math_lloyd:
 
-.. namespace:: oneapi::dal::kmeans::method
-.. struct:: lloyd
+Inference method: *Lloyd's*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Lloyd's inference method computes the :math:`y_j'` as index of the centroid
```
Comment (outoftardis) [inline source/elements/oneDAL/source/algorithms/clustering/kmeans.rst:113]:
```suggestion
Lloyd's inference method computes the :math:`y_j'` as an index of the centroid
```

File: source/elements/oneDAL/source/algorithms/clustering/kmeans.rst [hunk 23]
```diff
@@ -69,430 +95,102 @@ The steps (1) and (2) are performed until the following **stop condition**,
...
-Methods
--------
-::
+Inference
+---------
+Given the inference set :math:`X' = \{ x_1', \ldots, x_m' \}` of
+:math:`p`-dimensional feature vectors and the set :math:`C = \{ c_1, \ldots, c_k
+\}` of centroids produced at the training stage, the problem is to predict the
+index :math:`y_j' \in \{ 0, \ldots, k-1 \}`, :math:`1 \leq j \leq m`, of the
+centroid according to some method-defined rule.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/algorithms/clustering/kmeans.rst:107]:
```suggestion
centroid in accordance with a method-defined rule.
```

File: source/elements/oneDAL/source/algorithms/decomposition/pca.rst (hunk 2 above)
Comment (rlnx) [reply source/elements/oneDAL/source/algorithms/decomposition/pca.rst:126]: Not yet, will be part of other PR if time permits

File: source/elements/oneDAL/source/algorithms/decomposition/pca.rst (hunk 3 above)
Comment (rlnx) [reply source/elements/oneDAL/source/algorithms/decomposition/pca.rst:130]: Added

File: source/elements/oneDAL/include/oneapi/dal/kmeans.hpp [hunk 24]
```diff
@@ -0,0 +1,256 @@
...
+
+/// @tparam Task Tag-type that specifies type of the problem to solve. Can
+///              be :expr:`task::clustering`.
+template <typename Task = task::by_default>
+class model {
+public:
+   /// Creates a new instance of the class with the default property values.
+   model();
+
+   /// $k \\times p$ table with the cluster centroids. Each row of the
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans.hpp:68]:
```suggestion
   /// A $k \\times p$ table with the cluster centroids. Each row of the
```

File: source/elements/oneDAL/include/oneapi/dal/kmeans.hpp [hunk 25]
```diff
@@ -0,0 +1,256 @@
...
+///              be :expr:`task::clustering`.
+template <typename Task = task::by_default>
+class train_input {
+public:
+   /// Creates a new instance of the class with the given :literal:`data` and
+   /// :literal:`initial_centroids`
+   train_input(const table& data = table{},
+               const table& initial_centroids = table{});
+
+   /// $n \\times p$ table with the data to be clustered, where each row
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans.hpp:89]:
```suggestion
   /// An $n \\times p$ table with the data to be clustered, where each row
```

File: source/elements/oneDAL/include/oneapi/dal/kmeans.hpp [hunk 26]
```diff
@@ -0,0 +1,256 @@
...
+   /// stores one feature vector.
+   const table& get_data() const;
+   train_input& set_data(const table&);
+
+   /// $k \\times p$ table with the initial centroids, where each row
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans.hpp:94]:
```suggestion
   /// A $k \\times p$ table with the initial centroids, where each row
```

File: source/elements/oneDAL/include/oneapi/dal/kmeans.hpp [hunk 27]
```diff
@@ -0,0 +1,256 @@
...
+class train_result {
+public:
+   /// Creates a new instance of the class with the default property values.
+   train_result();
+
+   /// The trained K-means model
+   /// @remark default = model<Task>{}
+   const model<Task>& get_model() const;
+
+   /// $n \\times 1$ table with the labels $y_i$ assigned to the
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans.hpp:112]:
```suggestion
   /// An $n \\times 1$ table with the labels $y_i$ assigned to the
```

File: source/elements/oneDAL/include/oneapi/dal/kmeans.hpp [hunk 28]
```diff
@@ -0,0 +1,256 @@
...
+///              be :expr:`task::clustering`.
+template <typename Task = task::by_default>
+class infer_input {
+public:
+   /// Creates a new instance of the class with the given :literal:`model` and
+   /// :literal:`data`
+   infer_input(const model<Task>& m = model<Task>{},
+               const table& data = table{});
+
+   /// $n \\times p$ table with the data to be assigned to the clusters,
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans.hpp:167]:
```suggestion
   /// An $n \\times p$ table with the data to be assigned to the clusters,
```

File: source/elements/oneDAL/include/oneapi/dal/kmeans.hpp [hunk 29]
```diff
@@ -0,0 +1,256 @@
...
+
+/// @tparam Task Tag-type that specifies type of the problem to solve. Can
+///              be :expr:`task::clustering`.
+template <typename Task = task::by_default>
+class infer_result {
+public:
+   /// Creates a new instance of the class with the default property values.
+   infer_result();
+
+   /// $n \\times 1$ table with assignments labels to feature
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans.hpp:187]:
```suggestion
   /// An $n \\times 1$ table with assignments labels to feature
```

File: source/elements/oneDAL/include/oneapi/dal/pca.hpp [hunk 30]
```diff
@@ -0,0 +1,246 @@
...
+
+/// @tparam Task Tag-type that specifies type of the problem to solve. Can
+///              be :expr:`task::dim_reduction`.
+template <typename Task = task::by_default>
+class infer_result {
+public:
+   /// Creates a new instance of the class with the default property values.
+   infer_result();
+
+   /// $n \\times r$ table that contains data projected to the $r$
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/pca.hpp:187]:
```suggestion
   /// An $n \\times r$ table that contains data projected to the $r$
```

**Pull request merged**